### ISSUE TITLE
ND4J native slower than JBLAS for small matrices

### ISSUE THREAD

**Issue by nsalmon** [deeplearning4j/nd4j#1189]
It seems that using ND4J with native BLAS (e.g. OpenBLAS) is slower than using MTJ/netlib-java with JBLAS, when performing computation with small matrices.
I noticed this initially with a simple feedforward network.
I have boiled the benchmark to a simpler test : just checking matrix multiplication a matrix, for which i vary one dimension.
As dimension grows ND4J / native beats netlib-java, but for small matrice this is not true.

Code is available here :
https://github.com/nsalmon/nnetbench

Similar result on both Linux / Windows
BLAS : windows : OpenBLAS, Linux : not sure
DL4J/ND4J : 0.5.0

**Comment by treo** [comment]
You seem to be on to something here.
I've added your benchmarks on a branch of my JMH based suite, in order to make sure we don't run into the usual micro benchmark probles: https://github.com/treo/benchmarking_nd4j/tree/jblas-comparison

And those are my results: 
```
Benchmark                    Mode     Cnt     Score      Error  Units
MatrixMul.jblas_mult_50    sample  672820   291.965 ±    4.749  ns/op
MatrixMul.jblas_mult_100   sample  682634   527.705 ±    4.185  ns/op
MatrixMul.jblas_mult_200   sample  746667   910.276 ±   16.265  ns/op
MatrixMul.jblas_mult_300   sample  481726  1418.399 ±    8.814  ns/op
MatrixMul.jblas_mult_400   sample  750237  1752.942 ±   24.281  ns/op
MatrixMul.jblas_mult_500   sample  610629  2140.361 ±    8.233  ns/op
MatrixMul.jblas_mult_600   sample  515477  2532.502 ±    9.473  ns/op
MatrixMul.jblas_mult_700   sample  443655  2957.285 ±   11.689  ns/op
MatrixMul.jblas_mult_800   sample  390401  3396.029 ±   34.230  ns/op
MatrixMul.jblas_mult_900   sample  698368  3649.836 ±   12.422  ns/op
MatrixMul.jblas_mult_1000  sample  625307  4102.183 ±   12.547  ns/op
MatrixMul.nd4j_mult_50     sample  474364  2162.968 ± 1555.500  ns/op
MatrixMul.nd4j_mult_100    sample  541294  2158.111 ± 1320.987  ns/op
MatrixMul.nd4j_mult_200    sample  661643  2054.711 ±  805.799  ns/op
MatrixMul.nd4j_mult_300    sample  594561  3802.084 ± 2227.630  ns/op
MatrixMul.nd4j_mult_400    sample  537850  2679.986 ±  931.967  ns/op
MatrixMul.nd4j_mult_500    sample  484437  3020.257 ±  897.686  ns/op
MatrixMul.nd4j_mult_600    sample  508129  3477.612 ± 1166.689  ns/op
MatrixMul.nd4j_mult_700    sample  515286  3877.644 ± 1165.312  ns/op
MatrixMul.nd4j_mult_800    sample  499527  4068.887 ± 1193.907  ns/op
MatrixMul.nd4j_mult_900    sample  652886  3978.429 ±   21.687  ns/op
MatrixMul.nd4j_mult_1000   sample  609538  5079.889 ± 2180.397  ns/op
```

What we are seeing is that we have a huge variance on the timing of nd4j runs. And there seems to be a huge somewhat constant overhead, as the 10x sized version is only about 2.5x slower.

**Comment by raver119** [comment]
Not sure how jblas handles mult, hadn't got there yet, but mmul uses dup() inside. Can't it be the cause of the issue?

**Comment by AlexDBlack** [comment]
Result array is using createUninitialized: https://github.com/deeplearning4j/nd4j/blob/461f4f1fce039a9f250cfa7449b940e801eca232/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ndarray/BaseNDArray.java#L2465

I think dup is only used in places like this: https://github.com/deeplearning4j/nd4j/blob/461f4f1fce039a9f250cfa7449b940e801eca232/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ndarray/BaseNDArray.java#L2590
(Dup is necessary there to avoid overwriting the values we need to actually execute the mmul)

**Comment by saudet** [comment]
I modified the code slightly as below to run only the smallest configuration with ND4J and did some profiling.
```java
        final int numConfigurations = 1;
        final long numIter = 10000000;
        for (int i = 0; i < numConfigurations; i++) {
            int dim = (1 + i) * 100;
            long timeND4J = testND4J(dim, numIter);
        }
```

The hot spots look like this.
![hotspots](https://cloud.githubusercontent.com/assets/3328023/17456492/42ab0ad2-5c14-11e6-8667-9e70e4561ded.png)

So, there is a lot of time lost in allocating and deallocating native memory, obviously, but that doesn't account for all the difference. Since the buffers are small and could pretty much all fit in 256 KB memory, I think there's some cache trashing happening here. That would also account for the variance in execution time. The usual solution for that is to reuse buffers when possible, which is typically accomplished by calling the functions that take an output parameter instead of using the ones that return new objects, so let's try to do that...

**Comment by saudet** [comment]
Ah, doesn't help:
```java
    @Override
    public INDArray mmul(INDArray other, INDArray result) {
        return dup().mmuli(other, result);
    }
```

**Comment by saudet** [comment]
But if we use `mmuli()` directly then everything looks fine:
```java
    public static long testND4J(final int dim,
                                final long numIter) {
        INDArray input = Nd4j.rand(6, 1);
        INDArray w1 = Nd4j.rand(dim, 6);
        INDArray h1 = Nd4j.rand(dim, 1);

        long time = System.currentTimeMillis();
        for (int i = 0; i < numIter; i++) {
            w1.mmuli(input, h1);
        }

        return System.currentTimeMillis() - time;
    }
```

```
dim;time MTJ (ms);time ND4J (ms)
100;47;48
200;14;30
300;23;20
400;28;32
500;35;29
600;38;30
700;43;37
800;49;38
900;53;42
1000;61;48
...
```

**Comment by AlexDBlack** [comment]
re: return dup().mmuli(other, result);

I think a dup() is only necessary here when:
(a) other == result (same object)
(b) other and result share the same underlying DataBuffer
In both cases, we need to dup to avoid overwriting values that are being used in the mmul.

In all other cases, we should be able to remove the dup.

**Comment by treo** [comment]
Right, using mmuli with an already allocated target (i.e. doing it more like it is done with jblas), I get the following results (interleaved the jblas results from before for better comparability):
```
Benchmark                   Mode     Cnt     Score     Error  Units
MatrixMul.jblas_mult_50    sample  672820   291.965 ±    4.749  ns/op
MatrixMul.nd4j_mult_50    sample  712027   515.201 ±  10.701  ns/op

MatrixMul.jblas_mult_100   sample  682634   527.705 ±    4.185  ns/op
MatrixMul.nd4j_mult_100   sample  542749   680.958 ±  15.305  ns/op

MatrixMul.jblas_mult_200   sample  746667   910.276 ±   16.265  ns/op
MatrixMul.nd4j_mult_200   sample  772604   889.677 ±  25.272  ns/op

MatrixMul.jblas_mult_300   sample  481726  1418.399 ±    8.814  ns/op
MatrixMul.nd4j_mult_300   sample  590722  1159.232 ±  15.842  ns/op

MatrixMul.jblas_mult_400   sample  750237  1752.942 ±   24.281  ns/op
MatrixMul.nd4j_mult_400   sample  481340  1444.016 ±  48.033  ns/op

MatrixMul.jblas_mult_500   sample  610629  2140.361 ±    8.233  ns/op
MatrixMul.nd4j_mult_500   sample  417956  1753.220 ±  31.161  ns/op

MatrixMul.jblas_mult_600   sample  515477  2532.502 ±    9.473  ns/op
MatrixMul.nd4j_mult_600   sample  677684  1960.515 ± 116.781  ns/op

MatrixMul.jblas_mult_700   sample  443655  2957.285 ±   11.689  ns/op
MatrixMul.nd4j_mult_700   sample  586416  2266.992 ±  51.814  ns/op

MatrixMul.jblas_mult_800   sample  390401  3396.029 ±   34.230  ns/op
MatrixMul.nd4j_mult_800   sample  534458  2463.071 ±  19.203  ns/op

MatrixMul.jblas_mult_900   sample  698368  3649.836 ±   12.422  ns/op
MatrixMul.nd4j_mult_900   sample  476840  2767.260 ±  34.268  ns/op

MatrixMul.jblas_mult_1000  sample  625307  4102.183 ±   12.547  ns/op
MatrixMul.nd4j_mult_1000  sample  432102  3091.252 ±  33.472  ns/op
```

We can see that the variability has gone way down and that starting at 200 nd4j starts outperforming jblas.

**Comment by saudet** [comment]
Interesting. The difference is even larger when we tweak the code for MTJ to this:
```java
    public static long testMTJ(final int dim,
                               final long numIter) {

        Matrix w1 = randomMtjMatrix(dim, 6);
        DenseVector h1 = new DenseVector(dim);

        DenseVector inputVector = new DenseVector(6);
        for (int i = 0; i < 6; i++) {
            inputVector.set(i, 1.0);
        }

        long time = System.currentTimeMillis();
        for (int i = 0; i < numIter; i++) {
            w1.mult(inputVector, h1);
        }

        return System.currentTimeMillis() - time;
    }
```
Looks like the JVM is able to optimize that code globally better with JBLAS. It's probably inlining stuff and detecting some unnecessary computation given that the input data doesn't change and the output doesn't get used anywhere.

**Comment by saudet** [comment]
No, that's not it. There's a ~350 ns overhead per call and I have no idea where it's coming from. I've removed everything that could introduce the smallest delay. I've literally emptied out the call to `sgemv()`, but 99% of the time is still spent calling this function:
```cpp
JNIEXPORT void JNICALL Java_org_nd4j_nativeblas_Nd4jBlas_sgemv(JNIEnv* env, jclass cls) {
}
```
Something strange happening with JNI here...

**Comment by saudet** [comment]
It appears my profiler was playing tricks. Commenting out the call doesn't reduce actual execution time, but I get completely different hot spots:
![hotspots](https://cloud.githubusercontent.com/assets/3328023/17459870/4bff4012-5c87-11e6-9a3f-f1342d29e1b5.png)
So it looks like we'll need to reduce the allocation of new Java objects as well if we want to be able to compete with other libraries on small matrices.

**Comment by AlexDBlack** [comment]
Hm, interesting. So it's just the one GemvParameters object being allocation per mmul I think (assuming no copies are necessary). Only idea I have is to move the functionality currently in GemvParameters into BaseLevel2, saving the object allocation there. Could do the same for mmul too; also saves us a little bit on garbage collection at the expense of slightly messier code.

https://github.com/deeplearning4j/nd4j/blob/cf9fe733ace449bc797fb6b25879d8e5aa1683f1/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/blas/impl/BaseLevel2.java#L36

**Comment by saudet** [comment]
Actually, it's not the allocation itself that takes most of the time, but all the checks in the constructor. So I guess GemvParameters should be cached somehow. Not sure how since it can basically copy the whole array if necessary...

**Comment by agibsonccc** [comment]
Did this get fixed?

**Comment by saudet** [comment]
No, didn't get fixed, but it only affects small matrices...

**Comment by harperjiang** [comment]
I am using Nd4j to build customized LSTM networks and notice a similar problem. Besides allocateArray and deallocate, huge amount of time is consumed by BaseNDArrray.getDouble
![selection_006](https://cloud.githubusercontent.com/assets/1460389/23380636/40800800-fd01-11e6-8abb-cfc4685f07de.png)

I have two questions:
1. I checked the code to make sure I didn't invoke getDouble method in my code. How could I remove the "getDouble" hotspot?

2. I already tried to use in-place operations as much as possible to reduce native array allocation/deallocation. However, as the length of string in each batch varies, there are still many new arrays generated each round. My array size is 50 x [200~500] so I am not sure whether it falls in the "small matrices" scope. How could I reduce allocation time consumption in this case?
 
Any suggestion is greatly appreciated. Thanks!

**Comment by raver119** [comment]
Wow.

@harperjiang please file new issue, and bring all you have right there, your hotspots are very bad and very unusual, so i need to investigate them. 

Please include pom.xml, neural net configuration & optionally iterators (if you're using custom ones)

**Comment by raver119** [comment]
Stop.

Are you writing your OWN LSTM implementations? Or using ours?

**Comment by harperjiang** [comment]
@raver119  I am writing my own implementation directly using ND4J.

**Comment by raver119** [comment]
You're doing it int suboptimal way then. Working on getting scalars, creating lots of temp arrays etc. Check out how our implementation works inside.
